Inject constructor args

bajtos opened the pull request:
- Add a new Binding method `toClass(ctor)` to define a value that should be resolved by calling the constructor with arguments resolved using `@inject` annotation
 - Integrate this new API with Application and Controller concept

Connect to #142 
Requires #175 to be resolved first.

cc @bajtos @raymondfeng @ritch @superkhau

raymondfeng commented on packages/loopback/lib/context/index.ts:
```diff
@@ -25,6 +28,25 @@ export class Context {
     return binding;
   }
 
+  bindClass(key: string, ctor: new<T>(...args) => T): Binding {
```
Do we want to declare `new<T>(...args) => T` as a type?

raymondfeng replied:
Is it just a strongly-typed variant of `bind`? Can we still use `bind` to bind a key to a class?

raymondfeng commented on packages/loopback/lib/context/index.ts:
```diff
@@ -2,7 +2,10 @@
+export {inject};
 
 export class Context {
   private registry: Map<string, Binding>;
```
Should we support bindings using `class` or `interface` as keys?

bajtos replied:
> Should we support bindings using `class` or `interface` as keys?

Let's leave that out of scope of this pull request to keep it small and focused. 

I find the approach used in https://github.com/joelday/decoration-ioc (Visual Studio Code) as an interesting way for avoiding stringly-typed binding keys.

raymondfeng commented on packages/loopback/lib/context/inject.ts:
```diff
@@ -0,0 +1,44 @@
+ * @param bindingKey What binding to use in order to resolve the value
+ * of the annotated argument.
+ */
+export function inject(bindingKey: string) {
```
What about other flavors of `@inject`? For example:

- @inject(MyInterface | MyClass)
- @inject('my-key', filter) // Assuming multiple extensions can be registered under the same key

bajtos replied:
Let's leave that out of scope of this initial work please, see https://github.com/strongloop/loopback-next/pull/176#discussion_r112399332

raymondfeng commented on packages/loopback/lib/context/inject.ts:
```diff
@@ -0,0 +1,44 @@
+ */
+export function inject(bindingKey: string) {
+  return function markArgumentAsInjected(target: Object, propertyKey: string | symbol, parameterIndex: number) {
+    assert(parameterIndex != undefined, '@inject decorator can be used on function arguments only!');
```
I would like to use `@inject` for property injection as well.

ritch replied:
Just do this for now:

```ts
class MyClass {
  constructor(public @inject('foo') foo : string) { }
}
```

bajtos replied:
+1 considering property-based injection in the future. Let's leave that out of scope of this patch though.

raymondfeng commented on packages/loopback/lib/context/inject.ts:
```diff
@@ -0,0 +1,44 @@
+ * of the annotated argument.
+ */
+export function inject(bindingKey: string) {
+  return function markArgumentAsInjected(target: Object, propertyKey: string | symbol, parameterIndex: number) {
```
How can we make sure all arguments for the constructor are decorated with `@inject` so that the IoC container can instantiate the class?

bajtos replied:
I am afraid there is no generic and reliable way, considering that we want to support ES6 classes/constructors too.

I think our best option is to use `ctor.length` to determine the declared number of arguments and then verify that all of them were decorated.

raymondfeng commented on packages/loopback/test/acceptance/context-bindings/class-level-bindings.feature.md:
```diff
@@ -0,0 +1,33 @@
+   constructor(@inject('application.name') public appName: string) {
+   }
+ }
+ ctx.bindClass('controllers.info', InfoController);
```
What about the other styles?

- `ctx.bind('controllers.info').to(InfoController);`
- `ctx.bind('controllers.info').toClass(InfoController);`

ritch replied:
+1

ritch commented on packages/loopback/test/acceptance/context-bindings/class-level-bindings.ts:
```diff
@@ -0,0 +1,29 @@
+      constructor(@inject('application.name') public appName: string) {
+      }
+    }
+    ctx.bindClass('controllers.info', InfoController);
```
This doesn't really follow the convention of `bind()` + `to()`. If we really need `bindClass` I would expect it to be like this:

```ts
ctx.bind('controllers.info').toClass(InfoController);
// or
ctx.bind('controllers.info').toType(InfoController);
```

superkhau replied:
I agree with proposed chaining style.

bajtos commented:
> `bindClass(key: string, ctor: new<T>(...args) => T): Binding {`
> Is it just a strongly-typed variant of bind? Can we still use bind to bind a key to a class?

There are at least three different flavours of bound values that we need to support, and I am afraid JS/TS does not provide us enough information to distinguish between them automatically.

 1. Bind to a given static value (a primitive, an object/array, but also a function)
    ```ts
    ctx.bind('logger').to(console.log.bind(console));
    // ctx.get('logger')('Hello world');
    ```

 2. Bind to a getter function that will be called to retrieve the value to return
    ```ts
    ctx.bind('now').toDynamic(Date.now.bind(Date));
    // const now: number = ctx.get('now');
    ```

 3. Bind to a class constructor that will be called via `new` to retrieve the value to return
    ```ts
    ctx.bind('controllers.info').toClass(InfoController);
    // const controllerInstance: InfoController = ctx.get('controllers.info');
    ```

In particular:
 - 1 vs. 2 - there is no way how to tell whether a function argument was meant as a bound value or as a getter function
 - 2 vs. 3 - there is no way how to tell whether a function argument is a getter function (call it via `fn()`) or a class constructor (call it via `new Ctor()`).

bajtos commented:
```
ctx.bind('controllers.info').toClass(InfoController);
// or
ctx.bind('controllers.info').toType(InfoController);
```

I have considered that style too, but it requires all `Binding` instances to have a reference to their owning `Context`. Not a big deal, just something to be aware of. I'll rework the API per your suggestion.

bajtos commented:
I think I have addressed all of your review comments, PTAL again.

All comments are shown as outdated by GitHub because I have rebased the patch on top of the latest `master`.

bajtos commented:
Added one more commit integrating the DI implementation with application and controllers.

bajtos commented on packages/context/src/binding.ts:
```diff
@@ -3,16 +3,21 @@
 
-  constructor(private readonly _key: string, public isLocked: boolean = false) {}
+  // For bindings bound via toClass, this property contains the constructor function
+  public valueConstructor: Constructor<BoundValue>;
```
I need to access the constructor in order to retrieve `@api` metadata. I am not very happy with `.valueConstructor` property though, I think the main reason is because the property is used only by `toClass` bindings.

Any suggestions on how to improve this design?

bajtos commented on packages/context/src/context.ts:
```diff
@@ -20,11 +24,37 @@ export class Context {
+
+    for (let ix = 0; ix < fn.length; ix++) {
+      const bindingKey = injectedArgs[ix];
+      if (!bindingKey) {
```
This will throw on the first argument that's not injected, ignoring any other arguments that may be missing `@inject` decorator too. As a result, the user may need more than one iteration to fix the code.

We can detect all undecorated arguments upfront, but is the added complexity worth the cost? Thoughts?

superkhau replied:
IMO this is good enough for the initial impl, we can revisit if this turns into a major issue.

bajtos commented:
I have landed the PR to allow me to build more stuff on top of this base. I am happy to address any further comments in follow-up pull requests if necessary.
